docs/proposals/20201019-prometheus-rules-in-observatorium-api.md

```diff
@@ -0,0 +1,238 @@
+
+## Goals
+
+* Tenants can update their Prometheus recording rules without intervention from the operating team.
+* Observatorium must be able to scale horizontally scale with the number of tenants.
+
+## Non-Goals
+
+* Rate-limiting or tenant resource accounting.
+* Alerting rules. Although Prometheus `rule_files` can freely mix Recording and Alerting rules, alerting rules also require us to manage tenant-supplied routing and/or alerting configuration. This is out of scope for this design doc.
+* Logging rules. While Loki supports the same format rules as Prometheus, we will explicitly not consider supporting these in this proposal.
+
+## How
+
+We propose to solve the above problem by implementing a multi-tenant API that allows tenants to create, read, update and delete Prometheus recording rules.
+
+In Observatorium, [Thanos Rule](https://thanos.io/tip/components/rule.md/) evaluates recording rules. These are configured using a local file defined by Rule's `--rule.file` flag. `Rule` is configured as one of `Query`'s stores, so when rules are evaluated they are available to be queried.
+
+This leads to two problems we need to solve:
+1. How does `Rule` obtain tenant recording rules?
+2. How do we store tenant recording rules?
+
+### How does Rule get tenant recording rules
+
+For this step, we will leverage the [thanos-ruler-syncer](https://github.com/observatorium/thanos-rule-syncer).
+
+Periodically, this application does three things:
+1. Calls the Observatorium API to retrieve tenant recording rules.
+2. Writes these rules to a location defined by the `-file` flag.
+3. Asks Rule to reload its rules by calling the `/-/reload` endpoint.
+
+This will run as a sidecar to `Rule` and `-file` will be shared between the two containers.
+
+### How do we store tenant rules
+
+#### Service
+
+We will implement the rule storage backend as a separate service within Observatorium (i.e. not part of the API). This follows the established pattern of specifying backends in the API configuration and also maintains desirable properties of the API:
+* The API stays as a super-performant gateway that handles authentication and authorization but no application-specific logic.
+* Observatorium users are free to specify different storage backends, or no rule backend at all.
+
+The rule backend will be specified via the `--metrics.rules.endpoint` flag, and will route requests via the `/api/metrics/v1/{tenant}/rules` endpoint.
+
+NB: By scoping rules to the `/metrics` endpoint we are consciously creating room for future logging rules.
+
+Since we have explicitly de-scoped alerting rules from this proposal, the service will need to inspect the rules payload. If any rules are found to contain an `alert` field, this request should be rejected and the user returned a `400 Bad Request` error with a helpful message.
```

Comment:
I was wondering about this already, good job on answering questions before they can be asked :+1:. However, as an alternative, I was thinking whether accepting the mixed 'content' and giving the user warning about ignoring `alert`s would be more user-friendly. I guess it comes down whether we 'value' more user's convenience or user's level of awareness about what we are accepting (warning might as well be overlooked). I'm wondering about others' opinions.

Reply:
Hmm that is a very astute point :thinking: I'll add a section that dives into this decision in more detail and we can discuss there.